src/build.cc

```diff
@@ -474,6 +486,133 @@ bool Plan::CleanNode(DependencyScan* scan, Node* node, string* err) {
   return true;
 }
 
+namespace {
+bool EdgeCom(const Edge* lhs, const Edge* rhs) {
+  // Note: > to sort in decreasing order.
+  return lhs->critical_time() > rhs->critical_time();
+}
+}  // namespace
+
+void Plan::ComputePriorityList(BuildLog* build_log) {
+
+  for (vector<Node*>::iterator it=targets_.begin();it!=targets_.end();++it){
+      //remove duplicate targets
+      targets_.resize(find(it+1,targets_.end(),*it)-targets_.begin());
+  }
+
+  vector<Edge*> edges;
+  for (map<Edge*, bool>::iterator it = want_.begin(), end = want_.end();
+       it != end; ++it) {
+    if (!it->second)
+      continue;
+    Edge* e = it->first;
+    edges.push_back(e);
+  }
+  
+
+  // Critical path scheduling.
+  // 0. Assign costs to all edges, using:
+  // a) The time the edge needed last time, if available.
+  // b) The average time this edge type needed, if this edge hasn't run before.
+  // c) A fixed cost if this type of edge hasn't run before.
+  for (vector<Edge*>::iterator it = edges.begin(), end = edges.end(); it != end;
+       ++it) {
+    Edge* edge = *it;
+    if (edge->is_phony())
+      continue;  
+    BuildLog::LogEntry* entry =
+        build_log->LookupByOutput(edge->outputs_[0]->path());
+    if (!entry) {
+      edge->run_time_ms_ = 1;
+      continue;
+    }
+
+    int duration = entry->end_time - entry->start_time;  // XXX: + 1?
+    edge->run_time_ms_ = duration;
+    // XXX: also count on rule, and globally
+  }
+
+
+  // Dump graph to stdout for debugging / prototyping.
+  if (false) {
+    for (vector<Edge*>::iterator it = edges.begin(), end = edges.end();
+         it != end; ++it) {
+      Edge* edge = *it;
+      Node* input = edge->inputs_[0];
+      Node* output = edge->outputs_[0];
+      printf("%s %s %d\n", input->path().c_str(), output->path().c_str(),
+             edge->run_time_ms_);
+    }
+  }
+
+
+  uint64_t prev_top_crit=0;
+
+  for (vector<Node*>::reverse_iterator tit = targets_.rbegin(),
+                                       tend = targets_.rend();
+       tit != tend; ++tit) {
+    // 1. Use backflow algorithm to compute critical times for all nodes,
+    // starting from the destination nodes.
+    // XXX: ignores pools
+    queue<Edge*> edgesQ;
+
+    // Makes sure that each edge is added to the queue just once. This is needed
+    // for example if a binary is used to generate 50 source files, and all the
+    // source file cxx lines are added. Without this, the edge generating that
+    // binary would be added ot the queue 50 times.
+    set<Edge*> done;
+
+    Edge* first = (*tit)->in_edge();
+    if (!first){
+        continue;
+    } 
+    edgesQ.push(first);
+
+    first->set_critical_time(max(first->critical_time(),prev_top_crit));
```

Comment:
this should be corrected  as  :  first->set_critical_time(max(max(first->critical_time(),prev_top_crit),first->run_time_ms));